File: android/app/src/main/java/com/anydaf/data/api/FeedManager.kt
package com.anydaf.data.api

import android.content.Context
import android.util.Log
import com.anydaf.AnyDafApp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Mirrors iOS FeedManager.swift.
 *
 * Phase 1: Walk the SoundCloud RSS feed (paginated) for direct MP3 URLs.
 * Phase 2: Fetch each tractate's SoundCloud playlist to fill in any missing dafs
 *          (the RSS feed only covers the most recent ~1,200 episodes).
 *
 * Result is cached to filesDir/episode_index.json with a 7-day TTL.
 */
object FeedManager {
    private const val TAG = "FeedManager"
    private const val FEED_BASE = "https://feeds.soundcloud.com/users/soundcloud:users:958779193/sounds.rss"
    const val SOUNDCLOUD_CLIENT_ID = "1IzwHiVxAHeYKAMqN0IIGD3ZARgJy2kl"
    private const val SUPABASE_URL = "https://zewdazoijdpakugfvnzt.supabase.co"
    private const val SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inpld2Rhem9pamRwYWt1Z2Z2bnp0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzQ0NzIwODYsImV4cCI6MjA5MDA0ODA4Nn0.HJxIG18vEpt-exzoQwRLeXiKLAinWfBl7gMORKjxIz8"
    private const val CACHE_FILE = "episode_index.json"
    private const val CACHE_TIMESTAMP_KEY = "episodeIndexTimestamp"
    private const val CACHE_TTL_MS = 7L * 24 * 60 * 60 * 1000

    // Canonical tractate name → SoundCloud playlist ID
    private val tractatePlaylistIds = mapOf(
        "Berakhot"      to 1224453841L,
        "Shabbat"       to 1224957730L,
        "Eiruvin"       to 1224604675L,
        "Pesachim"      to 1223731237L,
        "Yoma"          to 1224408415L,
        "Sukkah"        to 1224961240L,
        "Beitzah"       to 1224467716L,
        "Rosh Hashanah" to 1225124800L,
        "Ta\u2019anit"  to 1947852215L,
        "Moed Katan"    to 1947706063L,
        "Chagigah"      to 1947633743L,
        "Yevamot"       to 1225156528L,
        "Ketubot"       to 1224649789L,
        "Nedarim"       to 1224705577L,
        "Nazir"         to 1950629151L,
        "Sotah"         to 1595841331L,
        "Gittin"        to 1224617542L,
        "Kiddushin"     to 1224719668L,
        "Bava Kamma"    to 1224873547L,
        "Bava Metzia"   to 1224692203L,
        "Bava Batra"    to 1224939157L,
        "Sanhedrin"     to 1225177738L,
        "Makkot"        to 1224421891L,
        "Shevuot"       to 1954367887L,
        "Avodah Zarah"  to 1224438616L,
        "Horayot"       to 1224645901L,
        "Zevachim"      to 1225250722L,
        "Menachot"      to 1950820791L,
        "Hullin"        to 1224735955L,
        "Bekhorot"      to 1224596788L,
        "Arakhin"       to 1224424696L,
        "Temurah"       to 1225194493L,
        "Meilah"        to 1224865387L,
        "Kinnim"        to 1954771503L,
        "Tamid"         to 1954771299L,
        "Middot"        to 1954771395L,
        "Niddah"        to 1225213678L,
    )

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val prefs get() =
        AnyDafApp.context.getSharedPreferences("feed_prefs", Context.MODE_PRIVATE)

    private val cacheFile get() =
        File(AnyDafApp.context.filesDir, CACHE_FILE)

    // tractate → daf → "soundcloud-track://ID" or direct MP3 URL
    private val _episodeIndex = MutableStateFlow<Map<String, Map<Int, String>>>(emptyMap())
    val episodeIndex: StateFlow<Map<String, Map<Int, String>>> = _episodeIndex.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    val hasIndex: Boolean get() = _episodeIndex.value.isNotEmpty()

    fun init() {
        loadFromCache()
    }

    /** Returns the audio URL string for a given tractate+daf, or null if not in the index. */
    fun audioUrl(tractate: String, daf: Int): String? =
        _episodeIndex.value[tractate]?.get(daf)

    /** Fetch only if cache is missing or older than 7 days.
     *  Tries Supabase first (fast single request); falls back to RSS crawl if unavailable. */
    suspend fun refreshIfNeeded() {
        val lastFetch = prefs.getLong(CACHE_TIMESTAMP_KEY, 0L)
        val age = System.currentTimeMillis() - lastFetch
        if (hasIndex && age <= CACHE_TTL_MS) return

        val supabaseIndex = fetchFromSupabase()
        if (supabaseIndex != null && supabaseIndex.isNotEmpty()) {
            _episodeIndex.value = supabaseIndex
            saveToCache(supabaseIndex)
            prefs.edit().putLong(CACHE_TIMESTAMP_KEY, System.currentTimeMillis()).apply()
            Log.d(TAG, "Loaded ${supabaseIndex.values.sumOf { it.size }} episodes from Supabase")
        } else {
            fetchAll()
        }
    }

    /** Fetch the full episode index from Supabase episode_audio table.
     *  Returns null if the request fails (caller should fall back to RSS crawl). */
    private suspend fun fetchFromSupabase(): Map<String, Map<Int, String>>? = withContext(Dispatchers.IO) {
        val url = "$SUPABASE_URL/rest/v1/episode_audio?select=tractate,daf,audio_url"
        val body = try {
            val req = Request.Builder()
                .url(url)
                .addHeader("apikey", SUPABASE_ANON_KEY)
                .addHeader("Authorization", "Bearer $SUPABASE_ANON_KEY")
                .build()
            client.newCall(req).execute().use { resp ->
                if (resp.code != 200) {
                    Log.w(TAG, "Supabase episode_audio HTTP ${resp.code}")
                    return@withContext null
                }
                resp.body?.string()
            }
        } catch (e: Exception) {
            Log.w(TAG, "Supabase episode_audio fetch failed", e)
            return@withContext null
        } ?: return@withContext null

        try {
            val arr = org.json.JSONArray(body)
            val index = mutableMapOf<String, MutableMap<Int, String>>()
            for (i in 0 until arr.length()) {
                val row      = arr.getJSONObject(i)
                val tractate = row.optString("tractate").takeIf { it.isNotEmpty() } ?: continue
                val daf      = row.optInt("daf", -1).takeIf { it > 0 } ?: continue
                val audioUrl = row.optString("audio_url").takeIf { it.isNotEmpty() } ?: continue
                index.getOrPut(tractate) { mutableMapOf() }[daf] = audioUrl
            }
            if (index.isEmpty()) null else index.mapValues { it.value.toMap() }
        } catch (e: Exception) {
            Log.w(TAG, "Supabase episode_audio JSON parse failed", e)
            null
        }
    }

    /** Force a full re-fetch from RSS + SoundCloud playlists. */
    suspend fun fetchAll() = withContext(Dispatchers.IO) {
        _isLoading.value = true
        val index = mutableMapOf<String, MutableMap<Int, String>>()

        // ── Phase 1: RSS feed ─────────────────────────────────────────────
        var nextUrl: String? = FEED_BASE
        var pageCount = 0
        while (nextUrl != null) {
            pageCount++
            Log.d(TAG, "RSS page $pageCount: $nextUrl")
            val xml = try {
                val req = Request.Builder().url(nextUrl).build()
                client.newCall(req).execute().use { it.body?.string() }
            } catch (e: Exception) {
                Log.e(TAG, "RSS fetch error", e); break
            } ?: break

            val parser = RssParser(xml)
            parser.parse()
            for (item in parser.items) {
                val tractate = item.tractate ?: continue
                val daf = item.daf ?: continue
                if (item.audioUrl.isEmpty()) continue
                index.getOrPut(tractate) { mutableMapOf() }.putIfAbsent(daf, item.audioUrl)
            }
            nextUrl = parser.nextPageUrl
        }

        // ── Phase 2: SoundCloud playlists (fill missing dafs) ─────────────
        Log.d(TAG, "Fetching ${tractatePlaylistIds.size} SoundCloud playlists in parallel")
        val playlistResults = coroutineScope {
            tractatePlaylistIds.map { (tractate, playlistId) ->
                async { fetchPlaylist(tractate, playlistId) }
            }.awaitAll()
        }

        for ((tractate, dafs) in playlistResults) {
            val existing = index.getOrPut(tractate) { mutableMapOf() }
            for ((daf, url) in dafs) {
                existing.putIfAbsent(daf, url) // don't overwrite RSS URLs
            }
        }

        _episodeIndex.value = index.mapValues { it.value.toMap() }
        saveToCache(index)
        prefs.edit().putLong(CACHE_TIMESTAMP_KEY, System.currentTimeMillis()).apply()
        Log.d(TAG, "fetchAll complete: ${_episodeIndex.value.values.sumOf { it.size }} total episodes")
        _isLoading.value = false
    }

    // ── SoundCloud playlist fetch ─────────────────────────────────────────

    private suspend fun fetchPlaylist(tractate: String, playlistId: Long): Pair<String, Map<Int, String>> {
        val dafs = mutableMapOf<Int, String>()
        val url = "https://api-v2.soundcloud.com/playlists/$playlistId?client_id=$SOUNDCLOUD_CLIENT_ID"
        val body = try {
            val req = Request.Builder().url(url).build()
            client.newCall(req).execute().use { it.body?.string() }
        } catch (e: Exception) {
            Log.e(TAG, "Playlist fetch error for $tractate", e)
            return tractate to dafs
        } ?: return tractate to dafs

        val json = try { JSONObject(body) } catch (e: Exception) { return tractate to dafs }
        val tracks = json.optJSONArray("tracks") ?: return tractate to dafs

        val fullTracks = mutableListOf<JSONObject>()
        val stubIds = mutableListOf<Long>()

        for (i in 0 until tracks.length()) {
            val track = tracks.getJSONObject(i)
            if (track.has("title")) fullTracks.add(track)
            else stubIds.add(track.optLong("id"))
        }

        // Batch-fetch stub tracks (SoundCloud only hydrates the first ~5 in a playlist)
        stubIds.chunked(50).forEach { batch ->
            val batchUrl = "https://api-v2.soundcloud.com/tracks" +
                    "?ids=${batch.joinToString(",")}&client_id=$SOUNDCLOUD_CLIENT_ID"
            try {
                val batchBody = client.newCall(Request.Builder().url(batchUrl).build())
                    .execute().use { it.body?.string() } ?: return@forEach
                val arr = JSONArray(batchBody)
                for (i in 0 until arr.length()) fullTracks.add(arr.getJSONObject(i))
            } catch (e: Exception) { /* skip batch */ }
        }

        for (track in fullTracks) {
            val title = track.optString("title")
            if (title.isEmpty()) continue
            val urn = track.optString("urn")
            if (urn.isEmpty()) continue
            val cleaned = title.replace(Regex("""\s*\(\d+\)\s*$"""), "").trim()
            val parts = cleaned.split(" ")
            if (parts.size < 2) continue
            val daf = parts.last().takeWhile { it.isDigit() }.toIntOrNull() ?: continue
            if (daf <= 0) continue
            val trackId = urn.split(":").lastOrNull()?.takeIf { it.isNotEmpty() } ?: continue
            dafs.putIfAbsent(daf, "soundcloud-track://$trackId")
        }

        return tractate to dafs
    }

    // ── Cache ─────────────────────────────────────────────────────────────

    private fun loadFromCache() {
        val file = cacheFile
        if (!file.exists()) return
        try {
            val root = JSONObject(file.readText())
            val index = mutableMapOf<String, MutableMap<Int, String>>()
            for (tractate in root.keys()) {
                // Migrate stale "Middos" key from before the rename to "Middot"
                val canonicalTractate = if (tractate == "Middos") "Middot" else tractate
                val dafObj = root.getJSONObject(tractate)
                val existing = index.getOrPut(canonicalTractate) { mutableMapOf() }
                for (key in dafObj.keys()) {
                    val dafNum = key.toIntOrNull() ?: continue
                    existing.putIfAbsent(dafNum, dafObj.getString(key))
                }
            }
            _episodeIndex.value = index.mapValues { it.value.toMap() }
            Log.d(TAG, "Loaded ${index.values.sumOf { it.size }} episodes from cache")
        } catch (e: Exception) {
            Log.e(TAG, "Cache load error", e)
        }
    }

    private fun saveToCache(index: Map<String, Map<Int, String>>) {
        try {
            val root = JSONObject()
            for ((tractate, dafs) in index) {
                val dafObj = JSONObject()
                for ((daf, url) in dafs) dafObj.put(daf.toString(), url)
                root.put(tractate, dafObj)
            }
            cacheFile.writeText(root.toString())
        } catch (e: Exception) {
            Log.e(TAG, "Cache save error", e)
        }
    }
}

File: AnyDaf/FeedManager.swift
import Foundation

@MainActor
class FeedManager: ObservableObject {
    @Published var isLoading = false
    @Published var loadingProgress = ""
    @Published var episodeCount: Int = 0

    // tractate name → daf number → audio URL string
    private(set) var episodeIndex: [String: [Int: String]] = [:]

    private let feedBase = "https://feeds.soundcloud.com/users/soundcloud:users:958779193/sounds.rss"
    static let soundcloudClientID = "1IzwHiVxAHeYKAMqN0IIGD3ZARgJy2kl"

    private let supabaseURL = "https://zewdazoijdpakugfvnzt.supabase.co"
    private let anonKey = Secrets.supabaseAnonKey

    // Canonical tractate name → SoundCloud playlist ID
    // Covers tractates that fall outside the RSS feed's rolling window
    private let tractatePlaylistIDs: [String: Int] = [
        "Berakhot":      1224453841,
        "Shabbat":       1224957730,
        "Eiruvin":       1224604675,
        "Pesachim":      1223731237,
        "Yoma":          1224408415,
        "Sukkah":        1224961240,
        "Beitzah":       1224467716,
        "Rosh Hashanah": 1225124800,
        "Ta\u{2019}anit": 1947852215,
        "Moed Katan":    1947706063,
        "Chagigah":      1947633743,
        "Yevamot":       1225156528,
        "Ketubot":       1224649789,
        "Nedarim":       1224705577,
        "Nazir":         1950629151,
        "Sotah":         1595841331,
        "Gittin":        1224617542,
        "Kiddushin":     1224719668,
        "Bava Kamma":    1224873547,
        "Bava Metzia":   1224692203,
        "Bava Batra":    1224939157,
        "Sanhedrin":     1225177738,
        "Makkot":        1224421891,
        "Shevuot":       1954367887,
        "Avodah Zarah":  1224438616,
        "Horayot":       1224645901,
        "Zevachim":      1225250722,
        "Menachot":      1950820791,
        "Hullin":        1224735955,
        "Bekhorot":      1224596788,
        "Arakhin":       1224424696,
        "Temurah":       1225194493,
        "Meilah":       1224865387,
        "Kinnim":        1954771503,
        "Tamid":         1954771299,
        "Middot":        1954771395,
        "Niddah":        1225213678,
    ]

    private let cacheURL: URL = {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        return docs.appendingPathComponent("episode_index.json")
    }()
    private let cacheTimestampKey = "episodeIndexTimestamp"

    var hasIndex: Bool { !episodeIndex.isEmpty }

    init() {
        loadFromCache()
    }

    func audioURL(tractate: String, daf: Int) -> URL? {
        guard let str = episodeIndex[tractate]?[daf] else { return nil }
        return URL(string: str)
    }

    // Force a fresh fetch regardless of cache age.
    // Always tries Supabase first; falls back to RSS+playlist crawl only if Supabase is unavailable.
    // Use this for the "Refresh Episodes" button and auto-retry on playback failure.
    func forceRefresh() async {
        isLoading = true
        loadingProgress = "Loading episodes…"
        defer {
            isLoading = false
            loadingProgress = ""
        }
        if let supabaseIndex = await fetchFromSupabase(), !supabaseIndex.isEmpty {
            episodeIndex = supabaseIndex
            episodeCount = supabaseIndex.values.reduce(0) { $0 + $1.count }
            saveToCache(supabaseIndex)
            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: cacheTimestampKey)
        } else {
            await fetchAll()
        }
    }

    // Fetch only if cache is missing, older than 7 days, or suspiciously small
    // (e.g. populated before the Supabase row-limit fix — full index is 2000+).
    // Tries Supabase first (fast single request); falls back to RSS crawl if unavailable.
    func refreshIfNeeded() async {
        let lastFetch = UserDefaults.standard.double(forKey: cacheTimestampKey)
        let age = Date().timeIntervalSince1970 - lastFetch
        guard !hasIndex || age > 7 * 24 * 3600 || episodeCount < 2000 else { return }

        if let supabaseIndex = await fetchFromSupabase(), !supabaseIndex.isEmpty {
            episodeIndex = supabaseIndex
            episodeCount = supabaseIndex.values.reduce(0) { $0 + $1.count }
            saveToCache(supabaseIndex)
            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: cacheTimestampKey)
        } else {
            await fetchAll()
        }
    }

    /// Fetch the full episode index from Supabase episode_audio table.
    /// Paginates in batches of 1000 to work around PostgREST's server-side max-rows cap.
    /// Returns nil if the request fails (caller should fall back to RSS crawl).
    private func fetchFromSupabase() async -> [String: [Int: String]]? {
        var index: [String: [Int: String]] = [:]
        let batchSize = 1000
        var offset = 0

        while true {
            let urlStr = "\(supabaseURL)/rest/v1/episode_audio"
                + "?select=tractate,daf,audio_url"
                + "&limit=\(batchSize)&offset=\(offset)"
                + "&order=tractate,daf"
            guard let url = URL(string: urlStr) else { return nil }
            var request = URLRequest(url: url)
            request.setValue(anonKey, forHTTPHeaderField: "apikey")
            request.setValue("Bearer \(anonKey)", forHTTPHeaderField: "Authorization")

            guard let (data, response) = try? await URLSession.shared.data(for: request),
                  (response as? HTTPURLResponse)?.statusCode == 200,
                  let rows = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
            else { return index.isEmpty ? nil : index }

            for row in rows {
                guard let tractate = row["tractate"] as? String,
                      let daf      = row["daf"]      as? Int,
                      let audioURL = row["audio_url"] as? String
                else { continue }
                if index[tractate] == nil { index[tractate] = [:] }
                index[tractate]![daf] = audioURL
            }

            if rows.count < batchSize { break }  // last page
            offset += batchSize
        }

        return index.isEmpty ? nil : index
    }

    func fetchAll() async {
        isLoading = true
        var index: [String: [Int: String]] = [:]

        // Phase 1: RSS feed (covers the most recent ~1,200 episodes)
        var nextURL: URL? = URL(string: feedBase)
        var pageCount = 0
        while let url = nextURL {
            pageCount += 1
            loadingProgress = "Loading episodes… (page \(pageCount))"
            guard let (data, _) = try? await URLSession.shared.data(from: url) else { break }
            let parser = RSSParser(data: data)
            parser.parse()
            for item in parser.items {
                guard let tractate = item.tractate, let daf = item.daf, !item.audioURL.isEmpty else { continue }
                if index[tractate] == nil { index[tractate] = [:] }
                if index[tractate]![daf] == nil { index[tractate]![daf] = item.audioURL }
            }
            nextURL = parser.nextPageURL
        }

        // Phase 2: SoundCloud playlist API — fills in dafs missing from the RSS
        loadingProgress = "Loading playlists…"
        let clientID = FeedManager.soundcloudClientID
        let playlistResults = await withTaskGroup(of: (String, [Int: String]).self) { group in
            for (tractate, playlistID) in tractatePlaylistIDs {
                group.addTask {
                    await FeedManager.fetchPlaylist(tractate: tractate,
                                                   playlistID: playlistID,
                                                   clientID: clientID)
                }
            }
            var collected: [(String, [Int: String])] = []
            for await result in group { collected.append(result) }
            return collected
        }

        for (tractate, dafs) in playlistResults {
            if index[tractate] == nil { index[tractate] = [:] }
            for (daf, url) in dafs {
                // Don't overwrite RSS URLs — they're direct MP3s (no resolution step needed)
                if index[tractate]![daf] == nil { index[tractate]![daf] = url }
            }
        }

        episodeIndex = index
        episodeCount = index.values.reduce(0) { $0 + $1.count }
        saveToCache(index)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: cacheTimestampKey)
        isLoading = false
        loadingProgress = ""
    }

    // Fetch all tracks from a SoundCloud playlist and return daf → soundcloud-track://ID URLs.
    // Static so it runs off the main actor, enabling true parallelism in the task group.
    private static func fetchPlaylist(tractate: String, playlistID: Int, clientID: String) async -> (String, [Int: String]) {
        var dafs: [Int: String] = [:]
        guard let url = URL(string: "https://api-v2.soundcloud.com/playlists/\(playlistID)?client_id=\(clientID)"),
              let (data, _) = try? await URLSession.shared.data(from: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tracks = json["tracks"] as? [[String: Any]]
        else { return (tractate, dafs) }

        // SoundCloud only hydrates the first ~5 tracks in a playlist response.
        // The rest come back as stubs (have an "id" but no "title"/"urn").
        // Collect stub IDs and batch-fetch full track data separately.
        var fullTracks: [[String: Any]] = []
        var stubIDs: [Int] = []
        for track in tracks {
            if track["title"] is String {
                fullTracks.append(track)
            } else if let id = track["id"] as? Int {
                stubIDs.append(id)
            }
        }

        // Fetch stubs in batches of 50 (SoundCloud API limit)
        let batchSize = 50
        var i = stubIDs.startIndex
        while i < stubIDs.endIndex {
            let batchEnd = stubIDs.index(i, offsetBy: batchSize, limitedBy: stubIDs.endIndex) ?? stubIDs.endIndex
            let batch = stubIDs[i..<batchEnd]
            let idsParam = batch.map(String.init).joined(separator: ",")
            if let batchURL = URL(string: "https://api-v2.soundcloud.com/tracks?ids=\(idsParam)&client_id=\(clientID)"),
               let (batchData, _) = try? await URLSession.shared.data(from: batchURL),
               let batchTracks = try? JSONSerialization.jsonObject(with: batchData) as? [[String: Any]] {
                fullTracks.append(contentsOf: batchTracks)
            }
            i = batchEnd
        }

        for track in fullTracks {
            guard let title = track["title"] as? String,
                  let urn = track["urn"] as? String else { continue }

            // Parse daf number (same logic as RSSParser.parseTitle)
            let cleaned = title.replacingOccurrences(of: #"\s*\(\d+\)\s*$"#, with: "",
                                                      options: .regularExpression)
            let parts = cleaned.split(separator: " ")
            guard parts.count >= 2,
                  let daf = Int(String(parts.last!).prefix(while: { $0.isNumber })),
                  daf > 0
            else { continue }

            // urn looks like "soundcloud:tracks:1004549059"
            guard let trackID = urn.split(separator: ":").last.map(String.init),
                  !trackID.isEmpty
            else { continue }

            if dafs[daf] == nil {
                dafs[daf] = "soundcloud-track://\(trackID)"
            }
        }

        return (tractate, dafs)
    }

    // MARK: - Cache

    private func loadFromCache() {
        guard let data = try? Data(contentsOf: cacheURL),
              var raw = try? JSONDecoder().decode([String: [String: String]].self, from: data)
        else { return }
        // Migrate stale "Middos" key from before the rename to "Middot"
        if let middos = raw["Middos"] {
            raw["Middot"] = (raw["Middot"] ?? [:]).merging(middos) { existing, _ in existing }
            raw.removeValue(forKey: "Middos")
        }
        episodeIndex = raw.mapValues { dafMap in
            Dictionary(uniqueKeysWithValues: dafMap.compactMap { key, url in
                guard let daf = Int(key) else { return nil }
                return (daf, url)
            })
        }
        episodeCount = episodeIndex.values.reduce(0) { $0 + $1.count }
    }

    private func saveToCache(_ index: [String: [Int: String]]) {
        let raw = index.mapValues { dafMap in
            Dictionary(uniqueKeysWithValues: dafMap.map { (String($0.key), $0.value) })
        }
        if let data = try? JSONEncoder().encode(raw) {
            try? data.write(to: cacheURL)
        }
    }
}
